### ISSUE TITLE
Create Policy for Datasets that will gets lots of memory at on-demand situation

### ISSUE THREAD

**Issue by kamisaberi** [kamisaberi/xtorch#717]
# Designing ImageDataFolder in C++: Preload vs. On-Demand Loading

## Key Considerations
- **Memory Usage**: Preloading all images consumes significant memory; on-demand loading minimizes it but may add latency.
- **Performance**: Preloading reduces runtime latency but increases startup time; on-demand loading spreads I/O across training.
- **Scalability**: Large datasets favor on-demand loading; preloading suits small datasets.
- **Use Case**: Depends on dataset size, hardware (RAM, disk speed), and data loader requirements.

### Option 1: Load All Images During Initialization
- **Description**: Read and store all image data (e.g., pixel buffers) in memory during `ImageDataFolder` construction.
- **Pros**:
  - Fast access during training (images in memory).
  - Simplifies data loader (no I/O during iteration).
- **Cons**:
  - High memory usage (e.g., 224x224 RGB image ≈ 150KB; 10,000 images ≈ 1.5GB).
  - Slow initialization for large datasets.
  - Infeasible for large datasets (e.g., ImageNet with 1.2M images).
- **Best For**:
  - Small datasets (<10,000 images) fitting in RAM.
  - Scenarios prioritizing training speed over startup time.

### Option 2: Store Image Names and Load During Data Loader
- **Description**: Store image file paths and labels during initialization, loading images on-demand in the data loader.
- **Pros**:
  - Low memory usage (only paths/metadata, ~KB per image).
  - Scales to large datasets (millions of images).
  - Fast initialization (only scans directory).
- **Cons**:
  - Slower access due to disk I/O per image.
  - Needs optimized data loader (prefetching, caching) to avoid bottlenecks.
- **Best For**:
  - Large datasets exceeding RAM.
  - Training with batching, shuffling, or augmentation.
  - Systems with fast storage (SSDs).

### Recommended Approach: Hybrid Strategy
- **Description**: Store image paths during initialization, use a cache in the data loader to preload a subset of images, and prefetch asynchronously.
- **Why?**:
  - Balances memory and performance.
  - Scales to large datasets while reducing I/O latency.
  - Leverages multi-core CPUs and fast storage.
- **Implementation Tips**:
  - Use a fixed-size LRU cache for recent images.
  - Load images in background threads.
  - Support batch prefetching.
- **Best For**: Most scenarios, especially deep learning with large datasets.

### Design Considerations
- **Dataset Size**: Preload for <10,000 images; use on-demand for larger datasets.
- **Hardware**:
  - **RAM**: Preloading feasible with 64GB+; otherwise, use on-demand.
  - **Storage**: SSDs reduce I/O latency for on-demand loading.
  - **CPU/GPU**: Multi-core CPUs enable parallel loading; GPUs need fast data pipelines.
- **Data Loader Requirements**:
  - Batching: Load images in groups.
  - Shuffling: Randomize access.
  - Augmentation: Apply transformations during loading.
- **Framework Integration**: Align with PyTorch/TensorFlow data loader patterns.

#### Option 1: Preload All Images
- **Structure**:
  ```cpp
  class ImageDataFolder {
  public:
      ImageDataFolder(const std::string& root_dir) {
          for (const auto& file : scanDirectory(root_dir)) {
              images_.push_back(loadImage(file.path));
              labels_.push_back(file.label);
          }
      }
      const Image& getImage(size_t index) const { return images_[index]; }
      size_t size() const { return images_.size(); }
  private:
      std::vector<Image> images_; // e.g., cv::Mat
      std::vector<int> labels_;
  };
  ```
- **Libraries**: OpenCV (`cv::imread`), libjpeg-turbo.
- **Challenges**:
  - Memory management for large buffers.
  - Slow initialization.
- **Optimizations**:
  - Compress images in memory (e.g., JPEG-encoded).
  - Use memory-mapped files.

#### Option 2: Load On-Demand
- **Structure**:
  ```cpp
  class ImageDataFolder {
  public:
      ImageDataFolder(const std::string& root_dir) {
          for (const auto& file : scanDirectory(root_dir)) {
              image_paths_.push_back(file.path);
              labels_.push_back(file.label);
          }
      }
      Image getImage(size_t index) const {
          return loadImage(image_paths_[index]);
      }
      size_t size() const { return image_paths_.size(); }
  private:
      std::vector<std::string> image_paths_;
      std::vector<int> labels_;
  };
  ```
- **Data Loader**:
  ```cpp
  class DataLoader {
  public:
      DataLoader(const ImageDataFolder& dataset, size_t batch_size)
          : dataset_(dataset), batch_size_(batch_size) {}
      std::vector<Image> nextBatch() {
          std::vector<Image> batch;
          for (size_t i = 0; i < batch_size_; ++i) {
              size_t idx = getNextIndex();
              batch.push_back(dataset_.getImage(idx));
          }
          return batch;
      }
  private:
      const ImageDataFolder& dataset_;
      size_t batch_size_;
  };
  ```
- **Challenges**:
  - I/O latency.
  - Thread safety for multi-threaded loading.
- **Optimizations**:
  - Prefetch next batch in background.
  - Cache frequent images.
  - Batch I/O operations.

#### Hybrid Approach
- **Structure**:
  ```cpp
  class ImageDataFolder {
  public:
      ImageDataFolder(const std::string& root_dir, size_t cache_size = 1000)
          : cache_(cache_size) {
          for (const auto& file : scanDirectory(root_dir)) {
              image_paths_.push_back(file.path);
              labels_.push_back(file.label);
          }
      }
      Image getImage(size_t index) {
          if (cache_.contains(index)) {
              return cache_.get(index);
          }
          Image img = loadImage(image_paths_[index]);
          cache_.put(index, img);
          return img;
      }
      void prefetch(const std::vector<size_t>& indices) {
          for (size_t idx : indices) {
              if (!cache_.contains(idx)) {
                  cache_.put(idx, loadImage(image_paths_[idx]));
              }
          }
      }
  private:
      std::vector<std::string> image_paths_;
      std::vector<int> labels_;
      LRUCache<size_t, Image> cache_;
  };
  ```
- **Features**:
  - LRU cache for recent images.
  - Asynchronous prefetching with threads.
  - Thread-safe cache access.
- **Libraries**:
  - OpenCV/stb_image for loading.
  - Boost/custom LRU cache.
  - `std::thread`/ThreadPool for async loading.

### Trade-offs
| **Approach**           | **Memory Usage** | **Initialization Time** | **Runtime Performance** | **Scalability** | **Complexity** |
|------------------------|------------------|-------------------------|-------------------------|-----------------|----------------|
| Preload All            | High             | Slow                    | Fast                    | Poor            | Low            |
| On-Demand              | Low              | Fast                    | Slow                    | Good            | Medium         |
| Hybrid (Cached)        | Moderate         | Fast                    | Moderate                | Good            | High           |

### Recommendations
- **Small Datasets (<10,000 images)**: Preload if RAM is sufficient.
- **Large Datasets (>10,000 images)**: Use hybrid approach with caching and prefetching.
- **Data Loader**: Support batching, shuffling, prefetching.
- **Optimizations**:
  - Use SSDs for faster I/O.
  - Multi-threaded loading.
  - Cache 100–1000 images.
  - Augment during loading.

### Example Workflow
1. **Initialization**: Scan directory, store paths/labels.
2. **Data Loader**: Fetch batches, shuffle, prefetch.
3. **Training**: Get batches, augment, feed to model.

### Resources
- **Libraries**:
  - [OpenCV](https://opencv.org/)
  - [stb_image](https://github.com/nothings/stb)
  - [Boost.Filesystem](https://www.boost.org/doc/libs/release/libs/filesystem/)
- **Guides**:
  - [PyTorch Data Loading](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html)
  - [TensorFlow Data Pipeline](https://www.tensorflow.org/guide/data_performance)

## Conclusion
The **hybrid approach** (store paths, cache with prefetching) is ideal for most cases, balancing memory, performance, and scalability. Use OpenCV for image processing, optimize with caching and multi-threading, and align with data loader best practices.

**Comment by kamisaberi** [comment]
# Sample Code:
```cpp


#include <opencv2/opencv.hpp>
#include <filesystem>
#include <vector>
#include <string>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <queue>
#include <unordered_map>
#include <list>
#include <stdexcept>

namespace fs = std::filesystem;

// Simple LRU Cache (thread-safe)
template<typename Key, typename Value>
class LRUCache {
public:
    LRUCache(size_t capacity) : capacity_(capacity) {}

    bool contains(Key key) {
        std::lock_guard<std::mutex> lock(mutex_);
        return cache_.find(key) != cache_.end();
    }

    Value get(Key key) {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = cache_.find(key);
        if (it == cache_.end()) {
            throw std::runtime_error("Key not found in cache");
        }
        // Move to front of LRU list
        lru_list_.splice(lru_list_.begin(), lru_list_, it->second.second);
        return it->second.first;
    }

    void put(Key key, const Value& value) {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = cache_.find(key);
        if (it != cache_.end()) {
            lru_list_.splice(lru_list_.begin(), lru_list_, it->second.second);
            it->second.first = value;
            return;
        }
        if (cache_.size() >= capacity_) {
            // Evict least recently used
            auto last = lru_list_.back();
            cache_.erase(last);
            lru_list_.pop_back();
        }
        lru_list_.push_front(key);
        cache_[key] = {value, lru_list_.begin()};
    }

private:
    size_t capacity_;
    std::mutex mutex_;
    std::unordered_map<Key, std::pair<Value, typename std::list<Key>::iterator>> cache_;
    std::list<Key> lru_list_;
};

class ImageDataFolder {
public:
    using Image = cv::Mat;

    ImageDataFolder(const std::string& root_dir, size_t cache_size = 1000)
        : cache_(cache_size), running_(true) {
        // Scan directory (assumes subdirs are class labels)
        for (const auto& class_dir : fs::directory_iterator(root_dir)) {
            if (!class_dir.is_directory()) continue;
            int label = std::stoi(class_dir.path().filename().string()); // e.g., "0", "1"
            for (const auto& file : fs::directory_iterator(class_dir)) {
                if (file.is_regular_file()) {
                    image_paths_.push_back(file.path().string());
                    labels_.push_back(label);
                }
            }
        }
        // Start cache-filling thread
        cache_thread_ = std::thread(&ImageDataFolder::cacheFiller, this);
    }

    ~ImageDataFolder() {
        {
            std::lock_guard<std::mutex> lock(queue_mutex_);
            running_ = false;
        }
        queue_cond_.notify_all();
        if (cache_thread_.joinable()) {
            cache_thread_.join();
        }
    }

    Image getImage(size_t index) {
        if (index >= image_paths_.size()) {
            throw std::out_of_range("Index out of range");
        }
        if (cache_.contains(index)) {
            return cache_.get(index);
        }
        Image img = loadImage(image_paths_[index]);
        cache_.put(index, img);
        return img;
    }

    void prefetch(const std::vector<size_t>& indices) {
        std::lock_guard<std::mutex> lock(queue_mutex_);
        for (size_t idx : indices) {
            if (idx < image_paths_.size()) {
                prefetch_queue_.push(idx);
            }
        }
        queue_cond_.notify_one();
    }

    size_t size() const { return image_paths_.size(); }
    int getLabel(size_t index) const { return labels_[index]; }

private:
    Image loadImage(const std::string& path) const {
        cv::Mat img = cv::imread(path, cv::IMREAD_COLOR);
        if (img.empty()) {
            throw std::runtime_error("Failed to load image: " + path);
        }
        return img;
    }

    void cacheFiller() {
        while (true) {
            size_t index;
            {
                std::unique_lock<std::mutex> lock(queue_mutex_);
                queue_cond_.wait(lock, [this] {
                    return !prefetch_queue_.empty() || !running_;
                });
                if (!running_ && prefetch_queue_.empty()) {
                    return;
                }
                if (prefetch_queue_.empty()) {
                    continue;
                }
                index = prefetch_queue_.front();
                prefetch_queue_.pop();
            }
            if (!cache_.contains(index)) {
                Image img = loadImage(image_paths_[index]);
                cache_.put(index, img);
            }
        }
    }

    std::vector<std::string> image_paths_;
    std::vector<int> labels_;
    LRUCache<size_t, Image> cache_;
    std::queue<size_t> prefetch_queue_;
    std::mutex queue_mutex_;
    std::condition_variable queue_cond_;
    std::thread cache_thread_;
    bool running_;
};

class DataLoader {
public:
    DataLoader(ImageDataFolder& dataset, size_t batch_size, bool shuffle = true)
        : dataset_(dataset), batch_size_(batch_size), shuffle_(shuffle), current_index_(0) {
        indices_.resize(dataset.size());
        for (size_t i = 0; i < indices_.size(); ++i) {
            indices_[i] = i;
        }
        if (shuffle_) {
            std::random_shuffle(indices_.begin(), indices_.end());
        }
    }

    std::vector<std::pair<ImageDataFolder::Image, int>> nextBatch() {
        std::vector<std::pair<ImageDataFolder::Image, int>> batch;
        std::vector<size_t> prefetch_indices;

        for (size_t i = 0; i < batch_size_ && current_index_ < indices_.size(); ++i) {
            size_t idx = indices_[current_index_];
            batch.emplace_back(dataset_.getImage(idx), dataset_.getLabel(idx));
            current_index_++;
        }

        // Prefetch next batch
        for (size_t i = current_index_; i < current_index_ + batch_size_ && i < indices_.size(); ++i) {
            prefetch_indices.push_back(indices_[i]);
        }
        if (!prefetch_indices.empty()) {
            dataset_.prefetch(prefetch_indices);
        }

        if (current_index_ >= indices_.size() && shuffle_) {
            current_index_ = 0;
            std::random_shuffle(indices_.begin(), indices_.end());
        }

        return batch;
    }

private:
    ImageDataFolder& dataset_;
    size_t batch_size_;
    bool shuffle_;
    std::vector<size_t> indices_;
    size_t current_index_;
};

// Example usage
int main() {
    ImageDataFolder dataset("/path/to/dataset", 1000); // Cache size: 1000 images
    DataLoader loader(dataset, 32, true); // Batch size: 32, shuffle: true

    for (int epoch = 0; epoch < 10; ++epoch) {
        while (true) {
            auto batch = loader.nextBatch();
            if (batch.empty()) break;
            // Process batch (e.g., feed to model)
            for (const auto& [image, label] : batch) {
                // Example: Print label (in practice, feed to model)
                std::cout << "Label: " << label << std::endl;
            }
        }
    }
    return 0;
}





```
